### PR TITLE
[WIP]: Extend gitDownloader to support new cache structure

### LINKED ISSUE #384: [LFX Term 3]: Full storage path for KCL third-party dependencies
# 1. Problem

- Dependencies from different registries may have the same name, and the current way kpm uses file names can cause conflicts that rely on duplicate names.
- The local saving path of the three-party library is case-sensitive, this has different effects on windows and unix systems.

# 2. Solution

In order to solve the problem, we redesigned the storage local system of the third-party dependencies with reference to rust

```
└── kpm
    ├── .kpm # all the configuration file for kpm client
    ├── git # all the kcl dependencies from git repo
    ├── oci # all the kcl dependencies from oci registry
```

## 2.1 Git

Under the subdir in `kpm/git`, the storage local system is

```
kpm/git
├── checkouts.   # checkout the specific version of git repository from cache bare repository 
│   ├── kcl-2a81898195a215f1
│   │   └── 33bb450. . # All the version of kcl package from git repository will be replaced with commit id
│   ├── kcl-578669463c900b87
│   │   └── 33bb450
└── db    # A bare git repository for cache git repo
    ├── kcl-2a81898195a215f1.      # <NAME>-<HASH> <NAME> is the name of git repo, 
    ├── kcl-578669463c900b87.   # <HASH> is calculated by the git full url.

``` 

NOTE: 

- `<NAME>-<HASH>` `<NAME>` is the name of git repo,  `<HASH>` is calculated by the git url. `<HASH>` is mainly used to distinguish between different git repositories, such as github and gitlab, and the reason for using a hash value is mainly to mask the inconsistency of the symbols in the url and local path.

## 2.2 Oci

Under the subdir in `kpm/oci`, the storage local system is

```
kpm/oci
├── cache # the cache for KCL dependencies tar
│   ├── ghcr.io-2a81898195a215f1    # <HOST>-<HASH> HOST is the name of oci registry,  <HASH> is calculated by the oci full url.
│   │   └── k8s_1.29.tar    # the tar for KCL dependencies
│   ├── docker.io-578669463c900b87
│   │   └── k8s_1.28.tar
└── src                                              	
│   ├── ghcr.io-2a81898195a215f1
│   │   └── k8s_1.29    # the KCL dependencies tar will untar here
│   ├── docker.io-578669463c900b87
│   │   └── k8s_1.28
``` 

# 4. Implementation

- The cache module for oci: https://github.com/kcl-lang/kpm/pull/502

### PR DESCRIPTION
**Pull request by Manoramsharma**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

fix #384 

#### 2. What is the scope of this PR (e.g. component or file name):

pkg/downloader/downloader.go
pkg/downloader/downloader_test.go
pkg/utils/utils.go

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [ ] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

We already have completed `CloneWithOpts` to support bare cloning , in this PR I have extended the `gitDownloader` to clone the dependency as bare repo to the cache path.

### CODE REVIEW

File: pkg/downloader/downloader.go [hunk 1]
```diff
@@ -356,16 +356,63 @@ func (d *GitDownloader) Download(opts DownloadOptions) error {
...
-		git.WithCommit(gitSource.Commit),
-		git.WithBranch(gitSource.Branch),
-		git.WithTag(gitSource.Tag),
-		git.WithRepoURL(gitSource.Url),
-		git.WithLocalPath(opts.LocalPath),
-	)
+	// If caching is enabled, clone the repo as a bare repository to the cache.
+	if opts.EnableCache {
+		// Check if the bare repository already exists in the cache.
+		if utils.DirExists(opts.CachePath) {
```
Comment (zong-zhe) [inline pkg/downloader/downloader.go:362]: Hi @Manoramsharma 😄 

 are some logical issues here that need to be considered:

1. Whether the bare repository exists.
2. Whether the corresponding `tag/commit/branch` exists in the bare repository.

Only when both conditions are met, that **the bare repository exists** and **that the corresponding tag/commit/branch exists in the repository** will the cache be hit, and the corresponding content will be cloned from the cache. Your current if statement only considers the first condition. The second condition needs to be added.

File: pkg/downloader/downloader.go [hunk 2]
```diff
@@ -356,16 +356,63 @@ func (d *GitDownloader) Download(opts DownloadOptions) error {
...
+			)
+			if err != nil {
+				return err
+			}
+		}
 
-	if err != nil {
-		return err
+		// Now checkout the specific version (commit/tag/branch) from the bare repository.
+		checkoutPath := filepath.Join(opts.LocalPath, "git", "checkouts", fmt.Sprintf("%s-%s", filepath.Base(gitSource.Url), utils.GenerateHash(gitSource.Url)))
```
Comment (zong-zhe) [inline pkg/downloader/downloader.go:384]: Hi @Manoramsharma 😄 

The hash is not only used here, the path to the bare repo should also be a hash value.

```shell
kpm/git
├── checkouts.   # checkout the specific version of git repository from cache bare repository 
│   ├── kcl-2a81898195a215f1
│   │   └── 33bb450. . # All the version of kcl package from git repository will be replaced with commit id
│   ├── kcl-578669463c900b87
│   │   └── 33bb450
└── db    # A bare git repository for cache git repo
    ├── kcl-2a81898195a215f1.      # <NAME>-<HASH> <NAME> is the name of git repo, 
    ├── kcl-578669463c900b87.   # <HASH> is calculated by the git full url.
```

Add more details, the hashed paths are default paths, that is, the user has not specified `localpath` and `cachepath` in `DownloadOptions`, we use the hash value to compute a default path, and if the user specifies paths, we use the specified paths directly.

Comment (Manoramsharma) [reply pkg/downloader/downloader.go:384]: Hi @zong-zhe can you help me with this one, how to use hash value to generate a default path? Currently I have covered only the part when the `localpath` and `cachepath` are defined by the user in `DownloadOptions`.

Where can I get the reference of using hash-value and genrate default storage path?

Comment (Manoramsharma) [reply pkg/downloader/downloader.go:384]: This `GenerateHash` function I have added in the utils should I move forward with this one? OR you have thought something different with hash-value and default path generation

**Pull request closed without merging**